marlowe-apps/src/Language/Marlowe/Runtime/App/Transact.hs

```diff
@@ -170,9 +181,36 @@ transactWithEvents event config key request =
             $ \case
               TxInfo{} -> pure ()
               response -> unexpected response
+          when (confirmSeconds > 0)
+            . withSubEvent subEvent (DynamicEventSelector "WaitAfterConfirm")
+            . const
+            $ liftIO . threadDelay . (confirmSeconds +)
+            =<< randomRIO (1_000_000, 2_000_000)
           pure contractId
 
 
+retry
+  :: T.Text
+  -> Event App r DynamicEventSelector f
+  -> [Int]
+  -> (Event App r DynamicEventSelector DynamicField -> App a)
+  -> App a
+retry name event [] action = withSubEvent event (DynamicEventSelector name) action
+retry name event (delay : delays) action =
+  withSubEvent event (DynamicEventSelector name)
+    $ \subEvent ->
+      ExceptT
```

Comment:
@jhbertra, could I have written this more succinctly, with less movement between `IO` and `ExceptT`?

Reply:
Would this work?

```hs
withSubEvent event (DynamicEventSelector name) \subEvent ->
  action subEvent `catchError` \message -> do
    addField subEvent $ ("failedAttempt" :: T.Text) ≔ message
    addField subEvent $ ("waitForRetry" :: T.Text) ≔ delay
    liftIO . threadDelay $ delay * 1_000_000
    retry name event delays action
```

Or if you want to un-nest the actions:

```hs
join $ withSubEvent event (DynamicEventSelector name) \subEvent ->
  (pure <$> action subEvent) `catchError` \message -> do
    addField subEvent $ ("failedAttempt" :: T.Text) ≔ message
    addField subEvent $ ("waitForRetry" :: T.Text) ≔ delay
    pure do
      liftIO . threadDelay $ delay * 1_000_000
      retry name event delays action
```

Reply:
Thanks so much. I'm not sure why I went to running the action instead of catching the error. 🤦

Reply:
Successfully tested 14a6290346e8fe1312b038bba6914c11f083f917 in [this run of `marlowe-scaling`](http://testing.marlowe.run/scaling/2023012435.log.gz) where options were set to provoke failures that would require retries. There were 52 retries.